lib/core/api/fetch-exception.mapper.ts

```diff
@@ -0,0 +1,110 @@
+} from "@lib/core/error/http";
+import { ServerErrorPayload } from "@lib/core/error/http/http-exception";
+
+/**
+ * Utility to map fetch errors to http-exceptions and provide support
+ * to add url to the exception message
+ *
+ * @param e - Error returned by fetch
+ * @param url - Url on which the call has been done
+ * @param method - Http method used for the call
+ * @param unexpectedErrorStatusCode - Default status code if exception is not supported
+ * @return HttpException or any supported concrete implementation
+ */
+export const mapFetchErrorToHttpException = (
+  e: Error | HttpException,
+  url?: string,
+  method?: string,
+  errorPayload?: ServerErrorPayload,
+  unexpectedErrorStatusCode = 500
+): HttpException => {
+  if (e instanceof HttpException) {
+    let status: number;
+    try {
+      status = e.status;
+    } catch (exception) {
+      status = unexpectedErrorStatusCode;
+    }
+    switch (status) {
+      case 400:
+        return new HttpBadRequest({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 401:
+        return new HttpUnauthorized({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 403:
+        return new HttpForbidden({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 404:
+        return new HttpNotFound({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 421:
+        return new HttpBadMapping({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 422:
+        return new HttpUnprocessableEntity({
+          url,
+          method,
+          previousError: e,
+          stackTrace: e.stack,
+          errorPayload,
+        });
+      case 405:
+        return new HttpMethodNotAllowed({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 408:
+        return new HttpRequestTimeout({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 500:
+        return new HttpInternalServerError({
+          url,
+          method,
+          previousError: e,
+          stackTrace: e.stack,
+          errorPayload,
+        });
+      case 501:
+        return new HttpNotImplemented({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 502:
+        return new HttpBadGateway({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      case 503:
+        return new HttpServiceUnavailable({
+          url,
+          method,
+          previousError: e,
+          stackTrace: e.stack,
+          errorPayload,
+        });
+      case 504:
+        return new HttpGatewayTimeout({ url, method, previousError: e, stackTrace: e.stack, errorPayload });
+      default:
+        return new HttpException({
+          status: status,
+          url,
+          method,
+          previousError: e,
+          stackTrace: e.stack,
+          errorPayload,
+        });
+    }
```

Comment:
It seems like having explicit errors for every status is a bit overkill - all of the sub-classes I've seen look exactly the same unless I missed something.

It'd be easier to maintain to have a root `HttpError` that has a `status`

Reply:
Agreed - a single error handler to handle errors and map them can be done in a single class and this is much easier to maintain later down the line. There is not much changing between status either; so a map can be done as a Set on the simple side, and with a error types on the complex side: